File: java/1498-number-of-subsequences-that-satisfy-the-given-sum-condition.java
class Solution {
    public int numSubseq(int[] nums, int target) {
        int mod = 1000000007;
        int n = nums.length;
        Arrays.sort(nums);

        int[] pow2 = new int[n];
        pow2[0] = 1;
        for (int i = 1; i < n; i++) {
            pow2[i] = (pow2[i - 1] * 2) % mod;
        }

        int left = 0, right = n - 1;
        int count = 0;

        while (left <= right) {
            if (nums[left] + nums[right] <= target) {
                count = (count + pow2[right - left]) % mod;
                left++;
            } else {
                right--;
            }
        }

        return count;
    }
}

File: kotlin/1498-number-of-subsequences-that-satisfy-the-given-sum-condition.kt
class Solution {
    fun numSubseq(nums: IntArray, target: Int): Int {
        nums.sort()
        val mod = 1000000000 + 7

        val pow = IntArray(nums.size)
        pow[0] = 1
        for (i in 0 until nums.size - 1)
            pow[i + 1] = pow[i] * 2 % mod

        var left = 0
        var right = nums.lastIndex
        var res = 0
        while (left <= right) {
            if (nums[left] + nums[right] > target) {
                right--
            } else {
                res = (res + pow[right - left]) % mod
                left++
            }
        }

        return res
    }
}
